File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.go
func BFS(y, x, l int, maps []string) [][]int {
	n := len(maps)
	m := len(maps[0])
	q := [][]int{{0, y, x}}
	visited := make([][]int, n)
	for i := range visited {
		visited[i] = make([]int, m)
	}

	for len(q) > 0 {
		cnt, nowY, nowX := q[0][0], q[0][1], q[0][2]
		q = q[1:]
		if l == 1 && string(maps[nowY][nowX]) == "L" {
			return [][]int{{cnt}, {nowY, nowX}}
		}
		if l == 0 && string(maps[nowY][nowX]) == "E" {
			return [][]int{{cnt}, {nowY, nowX}}
		}
		for _, next := range [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}} {
			nextY, nextX := next[0]+nowY, next[1]+nowX
			nowCnt := cnt + 1
			if 0 <= nextY && nextY < n && 0 <= nextX && nextX < m {
				if string(maps[nextY][nextX]) != "X" && visited[nextY][nextX] == 0 {
					q = append(q, []int{nowCnt, nextY, nextX})
					visited[nextY][nextX] = 1
				}
			}
		}
	}
	return [][]int{{-1}, {-1}}
}

func solution(maps []string) int {
	result := 0
	for y, width := range maps {
		for x, now := range width {
			if now == 'S' {
				bfsResult := BFS(y, x, 1, maps)
				if bfsResult[0][0] == -1 {
					return -1
				}
				result += bfsResult[0][0]
				bfsResult = BFS(bfsResult[1][0], bfsResult[1][1], 0, maps)
				if bfsResult[0][0] == -1 {
					return -1
				}
				result += bfsResult[0][0]
				return result
			}
		}
	}
	return -1
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    result = 0
    for y in range(len(maps)):
        for x in range(len(maps[y])):
            if maps[y][x] == 'S':
                count, now = BFS(y, x, maps, 1)
                if count == -1:
                    return -1
                result += count
                count, now = BFS(now[0], now[1], maps, 0)
                if count == -1:
                    return -1
                result += count
                return result


def BFS(y, x, maps, L):
    n = len(maps)
    m = len(maps[0])
    q = deque([[0,y,x]])
    visited = set()

    while q:
        cnt, now_y, now_x = q.popleft()
        if L == 1 and maps[now_y][now_x] == 'L':
            return (cnt, [now_y, now_x])
        if L == 0 and maps[now_y][now_x] == 'E':
            return (cnt, [now_y, now_x])
        for ny, nx in [[1,0],[-1,0],[0,1],[0,-1]]:
            next_y, next_x = ny+now_y, nx+now_x
            if 0 <= next_y < n and 0 <= next_x < m:
                if maps[next_y][next_x] != 'X' and (next_y, next_x) not in visited:
                    q.append([cnt+1, next_y, next_x])
                    visited.add((next_y, next_x))

    return (-1, -1)